### PR TITLE
GraphicsView and Microsoft.Maui.Graphics

### LINKED ISSUE #17: Microsoft.Maui.Graphics
MAUI includes the cross-platform [Microsoft.Maui.Graphics library](https://github.com/dotnet/Microsoft.Maui.Graphics), which is supported in MAUI apps, iOS, Android, Mac, WPF, UWP, WinForms, Tizen, and Linux. It includes the following features:

- Drawing canvas, supporting shapes, paths, shadows etc.
- Access fonts.
- Draw text
- Draw on bitmaps
- Create PDFs

The `GraphicsView` control, which is added in P5, can be used to consume this library.

### LINKED ISSUE #55: GraphicsView
MAUI includes a `GraphicsView` control, that can consume the types included in Microsoft.Maui.Graphics.

Document how to use this control.

Related to #17.

### PR DESCRIPTION
**Pull request by davidbritch**
Fixes #55 
Fixes #17

### CODE REVIEW

File: docs/user-interface/graphics/blendmodes.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
+---
+title: ".NET MAUI Graphics: Blend modes"
+description: "The .NET MAUI graphics library enables different compositing operations to be specified by the BlendMode property."
+ms.date: 07/30/2021
+---
+
+# .NET MAUI Graphics: Blend modes
```
Comment (adegeo) [inline docs/user-interface/graphics/blendmodes.md]: Don't forget to come back and alter these titles based on our discussions over the next day or so.

File: docs/user-interface/controls/graphicsview.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    {
+        public void Draw(ICanvas canvas, RectangleF dirtyRect)
+        {
+            // Drawing code goes here
+        }      
+    }
+}
+```
+
+The `Draw` method has `ICanvas` and `RectangleF` arguments. The `ICanvas` argument is the drawing canvas on which you draw graphical objects. The `RectangleF` argument is a `struct` that contains data about the size and location of the drawing canvas. For more information about drawing on an `ICanvas`, see [.NET MAUI Graphics: Draw objects](~/user-interface/graphics/draw.md).
```
Comment (adegeo) [inline docs/user-interface/controls/graphicsview.md]: There are a few instances of a site relative link `~/path/article` but we're not really supposed to use those unless required (such as if the publishing config imports a repo in here and you're linking to that cross-repo content). You can't click on them and get them working in vscode nor in github directly.

File: docs/user-interface/controls/graphicsview.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+title: ".NET MAUI GraphicsView"
+description: "The .NET MAUI GraphicsView is a graphics canvas on which 2D graphics can be drawn using the cross-platform Microsoft.Maui.Graphics library."
+ms.date: 07/28/2021
+---
+
+# .NET MAUI GraphicsView
+
+<!-- Sample link goes here -->
+
+The .NET Multi-platform App UI (.NET MAUI) `GraphicsView` is a graphics canvas on which 2D graphics can be drawn using the `Microsoft.Maui.Graphics` library. For more information about `Microsoft.Maui.Graphics`, see [.NET MAUI Graphics](~/user-interface/graphics/index.md).
```
Comment (adegeo) [inline docs/user-interface/controls/graphicsview.md]: Do we need to call out the library? Is it just included in the workload? I don't believe any of the winforms/wpf content calls out the library where types or concepts are held as it's just assumed the workload references these libraries for you, always.

Comment (davidbritch) [reply docs/user-interface/controls/graphicsview.md]: Maui.Graphics is both in the workload, and a library (available to use in .NET for iOS apps, . UWP apps, WPF apps etc.). Hence at the time of writing I called it out as a library. However my thinking is now this: as the docs are about MAUI, where you don't need to know that it's also a library, I should drop the library references. I could also add a note to the index page saying it's also available for single platform apps as a lib available from NuGet.

File: docs/user-interface/controls/graphicsview.md (hunk 2 above)
Comment (davidbritch) [reply docs/user-interface/controls/graphicsview.md]: To be cleaned up later by a script.

**Comment by davidbritch** [comment]
@jsuarezruiz 

The overview page definitely needs some work - I think we should drop the section on platform abstractions. The rest of the overview page is a bit random.

Also, is there anyway of forcing Maui.Graphics to use a specific graphics abstraction? e.g. force use of SkiaSharp on iOS rather than CoreGraphics

File: docs/user-interface/controls/graphicsview.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+In code, the `Drawable` property of the `GraphicsView` object can then be accessed, and the `ToImage` method called:
+
+```csharp
+using Microsoft.Maui.Graphics;
+...
+IImage image = graphicsDrawable.Drawable.ToImage(400, 500);
+```
+
+For information about image handling in `Microsoft.Maui.Graphics`, see [Images](~/user-interface/graphics/images.md).
```
Comment (jsuarezruiz) [inline docs/user-interface/controls/graphicsview.md:98]: Could we include information about how to invalidate the Canvas?. https://github.com/dotnet/maui/pull/3142
I guess could be a typical question.

File: docs/user-interface/graphics/colors.md [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+- `Color.FromRgba` from a `string`-based hexadecimal value in the form "#RRGGBBAA" or "#RRGGBB" or "#RGBA" or "#RGB", where each letter corresponds to a hexadecimal digit for the alpha, red, green, and blue channels.
+- `Color.FromHsla` from `float` HSLA values.
+- `Color.FromHsla` from `double` HSLA values.
+- `Color.FromHsv` from `float` HSV values that range from 0 to 1.
+- `Color.FromHsv` from `int` HSV values that range from 0 to 255.
+- `Color.FromHsva` from `float` HSVA values.
+- `Color.FromHsva` from `int` HSV values.
+- `Color.FromInt` from an `int` value calculated as (B + 256 \* (G + 256 \* (R + 256 \* A))).
+- `Color.FromUint` from a `uint` value calculated as (B + 256 \* (G + 256 \* (R + 256 \* A))).
+- `Color.FromArgb` from a `string`-based hexadecimal value in the form "#AARRGGBB" or "#RRGGBB" or "#ARGB" or "RGB", where each letter corresponds to a hexadecimal digit for the alpha, red, green, and blue channels.
```
Comment (jsuarezruiz) [inline docs/user-interface/graphics/colors.md:37]: I don't know if it is worth highlighting, but whoever is migrating from Xamarin.Forms should use this method to replace `Color.FromHex`

File: docs/user-interface/graphics/index.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+- `Microsoft.Maui.Graphics` functionality must be consumed on a drawing canvas, enables performant graphics to be drawn, and provides a convenient approach for writing graphics-based controls. For example, a control that replicates the GitHub contribution profile can be more easily implemented using `Microsoft.Maui.Graphics` than by using .NET MAUI shapes.
+- .NET MAUI shapes can be consumed directly on a page, and brushes can be consumed by all controls. This functionality is provided to help you produce an attractive UI.
+
+## Platform abstractions
+
+The following table lists the graphics abstractions that are supported on each platform:
+
+| Platform | Graphics abstractions |
+| -- | -- |
+| .NET MAUI | Platform support as shown per platform below. |
```
Comment (jsuarezruiz) [inline docs/user-interface/graphics/index.md:28]: I think that we will have a Core library maintained with support for the platforms officially supported by .NET MAUI and another Community library with the rest of the platforms. @Redth could we confirm this? If so, just mention in documentation the officially supported platforms?

File: docs/user-interface/graphics/index.md [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+| Tizen | SkiaSharp |
+| Linux | SkiaSharp |
+
+By default, .NET MAUI uses the native graphics capabilities of each platform.
+
+For examples of using different graphics abstractions, see [Microsoft.Maui.Graphics samples](https://github.com/dotnet/Microsoft.Maui.Graphics/tree/main/samples).
+
+## Drawing canvas
+
+In .NET MAUI, the `GraphicsView` enables consumption of the `Microsoft.Maui.Graphics` functionalit, via a drawing canvas that's exposed as an `ICanvas` object. For more information about the `GraphicsView`, see [GraphicsView](~/user-interface/controls/graphicsview.md).
```
Comment (jsuarezruiz) [inline docs/user-interface/graphics/index.md]: functionalit = functionality

File: docs/user-interface/graphics/colors.md [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+# Colors
+
+<!-- Sample link goes here -->
+
+The `Color` class, in the `Microsoft.Maui.Graphics` namespace, lets you specify colors as Red-Green-Blue (RGB) values, Hue-Saturation-Luminosity (HSL) values, Hue-Saturation-Value (HSV) values, or with a color name. An Alpha channel is also available to indicate transparency.
+
+`Color` objects can be created with `Color` constructors, which can be used to specify a gray shade, an RGB value, or an RGB value with transparency. In all cases, arguments are `float` values ranging from 0 to 1.
+
+> [!NOTE]
+> The default `Color` constructor creates a black `Color` object.
```
Comment (jsuarezruiz) [inline docs/user-interface/graphics/colors.md:18]: Also, Color.Default (from Xamarin.Forms) not exists here. The comparison to detect if a color property has not been initialized is:

`myColor == null`

Instead:

`myColor == Color.Default`

I am thinking in devs coming from Xamarin.Forms, this kind of details could help in migrations, etc.

File: docs/user-interface/graphics/index.md [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+- `BlendMode`, of type `BlendMode`, defines the blend mode, which determines what happens when an object is rendered on top of an existing object.
+- `DisplayScale`, of type `float`, represents the scaling factor to scale the UI by on a canvas.
+- `FillColor`, of type `Color`, indicates the color used to paint an object's interior.
+- `FontColor`, of type `Color`, specifies the font color when drawing text.
+- `FontName`, of type `string`, defines the font when drawing text.
+- `FontSize`, of type `float`, defines the size of the font when drawing text.
+- `MiterLimit`, of type `float`, specifies the limit of the miter length of line joins in an object.
+- `StrokeColor`, of type `Color`, indicates the color used to paint an object's outline.
+- `StrokeDashPattern`, of type `float[]`, specifies the pattern of dashes and gaps that are used to outline an object.
+- `StrokeLineCap`, of type `LineCap`, describes the shape at the start and end of a line.
```
Comment (jsuarezruiz) [inline docs/user-interface/graphics/index.md:59]: To align properties and avoid duplicates types between Graphics and Maui, the type of some properties could change.

File: docs/user-interface/graphics/colors.md (hunk 5 above)
Comment (davidbritch) [reply docs/user-interface/graphics/colors.md:37]: I won't highlight this here, but there will be a migration doc that lists all the changes Xamarin.Forms users need to know about.

File: docs/user-interface/graphics/colors.md (hunk 8 above)
Comment (davidbritch) [reply docs/user-interface/graphics/colors.md:18]: I'll save this for the migration doc.

**Comment by davidbritch** [comment]
Replaced with #241

**Pull request closed without merging**